### PR TITLE
Accessing dtype via `.dtype` attribute as opposed to `encoding` hash map

### LINKED ISSUE #45: Change how we access dtype
Currently we are accessing dtype from the `da.encoding` dict. We should be able to use `da.dtype` instead, which is less likely to be altered by various data processing steps.

https://github.com/xpublish-community/xpublish-opendap/blob/fc02c729bc619ee76dd3ece5ffa4333f2b335fa9/xpublish_opendap/dap_xarray.py#L78

xref: https://github.com/xpublish-community/xpublish/discussions/243

### PR DESCRIPTION
**Pull request by xaviernogueira**
Closes #45 

A simple fix to use the `xarray.Dataset(...).dtpye` attribute to read dtype as opposed to accessing the `encoding` hashmap, which may be wiped or empty in certain conditions.

### CODE REVIEW

**Comment by xaviernogueira** [comment]
@abkfenris seems that 3.9 compatibility isn't working -> I'll see if I can get this resolved.

**Comment by abkfenris** [comment]
I'd check if the Xarray versions are the same. I believe there has been a lot of refactoring around `Variable`s recently.

**Comment by abkfenris** [comment]
I've used https://github.com/mxschmitt/action-tmate to get access to a running Actions environment for debugging. I'd suggest making a new workflow with a single python environment rather than trying to run that across the matrix though.

**Comment by xaviernogueira** [comment]
Good tip! I'll use that next time. 

As of now, I think I figured it out. There are two issues:
* In ubuntu-3.9, the `xarray.DataArray.dtype` is returned as ">f4" (string), as opposed to a `np.dtype` object.
* np.dtype('<f4')` evaluates to `np.dtype('float32')`, therefore my initial solution was to just use `np.dtype(da.dtype)` to evaluate to one of the core numpy dtypes (that are keys in our `dap_dtype_dict`). That said, this does not work because np.dtype('**>**f4') evaluates to itself.

The solution (think) is to leave the addition where we use `np.dtype()` to evaluate any string `xarray.DataArray.dtype`, but to also add additional dtypes to `dap_dtypes_dict`. @abkfenris 


Basically, for  One can 
As of now I was able to use the logging commits and see that it's simply that we don't have `np.dtype('>f4')` in `dap_dict_dtype`. This originally led to an attribute error during the logging of a warning as `da.name` did not exist. Now it gets past that but fails the final assert statements as the default `dap.String` dtype is used.

**Comment by abkfenris** [comment]
Great, you're definitely welcome to add to the `dap_dtypes_dict`. It's definitely not exhaustive as I built it out by hand with some of my data and the tutorials dataset.

**Comment by xaviernogueira** [comment]
@abkfenris  Well that was more involved than I expected! All works now.

**Two Issues:** 
* Some `xr.DataArray.encoding` dictionaries are empty or can be wiped.
* I discovered that the issue with Python 3.9 arose only from `xr.conventions.encode_cf_variable` which would convert dtype to `>f4`, which is backward encoding from `<f4` == `float32`.

**Implemented Changes:**
* Dtype is read via `.dtype` attr instead of the encodings dict.
* Fixed the type-hinting of `dap_dtype()` to reflect that `da` can/will be both `xr.DataArray` and `xr.Variable` types.
* Took away `da.name` from the KeyError handling `dap_dtype()` as `xr.Variable` inputs don't have a name attr, and this causes an AttributeError.
* In `dap_dimensions`, if the encoding is done backwards ">", we re-cast via `astype` to the normal version by simply replacing > with < in the dtype string. So no need to update the `dap_dtypes_dict`.

**Comment by xaviernogueira** [comment]
Had to squeeze in some minor edits. Merge ready!

**Review by abkfenris — APPROVED** [review]
Thank you!
